### PR TITLE
feat: `riscv-interrupt-{m,s}` calling conventions

### PR DESCRIPTION
**Pull request by sethp**
Similar to prior support added for the mips430, avr, and x86 targets this change implements the rough equivalent of clang's [`__attribute__((interrupt))`][clang-attr] for riscv targets, enabling e.g.

```rust
static mut CNT: usize = 0;

pub extern "riscv-interrupt-m" fn isr_m() {
    unsafe {
        CNT += 1;
    }
}
```

to produce highly effective assembly like:

```asm
pub extern "riscv-interrupt-m" fn isr_m() {
420003a0:       1141                    addi    sp,sp,-16
    unsafe {
        CNT += 1;
420003a2:       c62a                    sw      a0,12(sp)
420003a4:       c42e                    sw      a1,8(sp)
420003a6:       3fc80537                lui     a0,0x3fc80
420003aa:       63c52583                lw      a1,1596(a0) # 3fc8063c <_ZN12esp_riscv_rt3CNT17hcec3e3a214887d53E.0>
420003ae:       0585                    addi    a1,a1,1
420003b0:       62b52e23                sw      a1,1596(a0)
    }
}
420003b4:       4532                    lw      a0,12(sp)
420003b6:       45a2                    lw      a1,8(sp)
420003b8:       0141                    addi    sp,sp,16
420003ba:       30200073                mret
```

(disassembly via `riscv64-unknown-elf-objdump -C -S --disassemble ./esp32c3-hal/target/riscv32imc-unknown-none-elf/release/examples/gpio_interrupt`)

This outcome is superior to hand-coded interrupt routines which, lacking visibility into any non-assembly body of the interrupt handler, have to be very conservative and save the [entire CPU state to the stack frame][full-frame-save]. By instead asking LLVM to only save the registers that it uses, we defer the decision to the tool with the best context: it can more accurately account for the cost of spills if it knows that every additional register used is already at the cost of an implicit spill.

At the LLVM level, this is apparently [implemented by] marking every register as "[callee-save]," matching the semantics of an interrupt handler nicely (it has to leave the CPU state just as it found it after its `{m|s}ret`).

This approach is not suitable for every interrupt handler, as it makes no attempt to e.g. save the state in a user-accessible stack frame. For a full discussion of those challenges and tradeoffs, please refer to [the interrupt calling conventions RFC][rfc].

Inside rustc, this implementation differs from prior art because LLVM does not expose the "all-saved" function flavor as a calling convention directly, instead preferring to use an attribute that allows for differentiating between "machine-mode" and "superivsor-mode" interrupts.

Finally, some effort has been made to guide those who may not yet be aware of the differences between machine-mode and supervisor-mode interrupts as to why no `riscv-interrupt` calling convention is exposed through rustc, and similarly for why `riscv-interrupt-u` makes no appearance (as it would complicate future LLVM upgrades).

[clang-attr]: https://clang.llvm.org/docs/AttributeReference.html#interrupt-risc-v
[full-frame-save]: https://github.com/esp-rs/esp-riscv-rt/blob/9281af2ecffe13e40992917316f36920c26acaf3/src/lib.rs#L440-L469
[implemented by]: https://github.com/llvm/llvm-project/blob/b7fb2a3fec7c187d58a6d338ab512d9173bca987/llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp#L61-L67
[callee-save]: https://github.com/llvm/llvm-project/blob/973f1fe7a8591c7af148e573491ab68cc15b6ecf/llvm/lib/Target/RISCV/RISCVCallingConv.td#L30-L37
[rfc]: https://github.com/rust-lang/rfcs/pull/3246

### CODE REVIEW

**Comment by sethp** [comment]
So the riscv UI test failures have taken some debugging. It seems that these are the first riscv-specific UI tests, which have been subtly broken since #110884: on any version of LLVM that does not recognize that feature (i.e. any before https://github.com/rust-lang/llvm-project/commit/da7c7123d9acfa9b699e85dae733318e6ec13114, which appears to have landed in ~15.0.0), the act of [filtering down] the total list of features [known to rustc](https://github.com/rust-lang/rust/blob/c373194cb6d882dc455a588bcc29c92a96b50252/compiler/rustc_codegen_ssa/src/target_features.rs#L244) attempts to apply each feature to a [short-lived feature set](https://github.com/llvm/llvm-project/blob/25a4a9360e1dedb572dc007cf8cc97dfc3efdeac/llvm/lib/MC/MCSubtargetInfo.cpp#L302), causing LLVM to emit this error as a [side effect](https://github.com/llvm/llvm-project/blob/25a4a9360e1dedb572dc007cf8cc97dfc3efdeac/llvm/lib/MC/MCSubtargetInfo.cpp#L82-L83):

```
+ '+unaligned-scalar-mem' is not a recognized feature for this target (ignoring feature)
```

Since this goes directly to stderr, the UI tests pick it up and observe "hey, on LLVM 14 there's this extra output that's not in the .stderr file": and on versions that do support the feature, no such message is printed. (I suspect, but have not confirmed, that this means every feature listed in that file is supported across every tested version of LLVM, except apparently for this one).

As a potential fix, in https://github.com/rust-lang/rust/pull/111891/commits/70b2188c36ea4730a22b219617cefab82e2e7baa I've modified the C++ side of the binding to search the feature table directly rather than delegating to the side-effecting class member, but this requires access to the feature table. Accessors were added to the Rust fork in https://github.com/rust-lang/llvm-project/commit/da7c7123d9acfa9b699e85dae733318e6ec13114 , but the failing test is being built against Ubuntu's system LLVM that lacks any rust-specific patches, so that accessor doesn't exist. Instead, I've had to resort to, uh, cleverness to access the private field in a way that the internet claims ought to be portable across LLVM and language versions, and C++ compilers. Feedback gladly accepted on where else there might be some leverage here that I'm missing, either within the test suite or elsewhere.

[filtering down]: https://github.com/rust-lang/rust/blob/c373194cb6d882dc455a588bcc29c92a96b50252/compiler/rustc_codegen_llvm/src/llvm_util.rs#L301

**Comment by jackh726** [comment]
This likely needs some lang and/or compiler team okay

**Comment by sethp** [comment]
Thanks for taking a look, @jackh726! Is this something the lang and/or compiler teams are going to see, or is there an action I should be taking to discuss it with them?

**Comment by jackh726** [comment]
I'm going to start by tagging @rust-lang/lang. I guess adding a new calling convention needs at least someone on lang to okay (it's behind a feature gate, so likely doesn't need full FCP). For compiler, not sure - maybe an MCP is warranted, but probably not.

For lang team: Is it okay to add a feature gate for new `risvc-interrupt-{m,s}` calling conventions?

**Comment by jackh726** [comment]
Lang nominating. See above comment.

**Comment by nikomatsakis** [comment]
@jackh726 this sounds like an [experimental feature gate](https://lang-team.rust-lang.org/how_to/experiment.html), although I think it would ultimately be covered under https://github.com/rust-lang/rfcs/pull/3246. 

@tmandry can serve as liaison.

File: src/ci/run.sh [hunk 1]
```diff
@@ -8,7 +8,7 @@ fi
 
 if [ "$NO_CHANGE_USER" = "" ]; then
   if [ "$LOCAL_USER_ID" != "" ]; then
-    useradd --shell /bin/bash -u $LOCAL_USER_ID -o -c "" -m user
+    id -u $LOCAL_USER_ID &>/dev/null || useradd --shell /bin/bash -u $LOCAL_USER_ID -o -c "" -m user
```
Comment (jackh726) [inline src/ci/run.sh:11]: This change doesn't seem like it should be in this PR

**Review by jackh726 — CHANGES_REQUESTED** [review]

File: src/ci/run.sh (hunk 1 above)
Comment (sethp) [reply src/ci/run.sh:11]: Thanks for the feedback! I'm not quite sure whether you're asking for context, or telling me to revert this change? 

I remember this change being intentional as part of [605858b](https://github.com/rust-lang/rust/commit/605858b83f8c4cfc889afca0374ac4c078033f5e), which I believe I could most effectively test using the CI infrastructure's docker containers for acquiring a ubuntu-flavored LLVM 14. IIRC every part of the CI process was idempotent/well-suited to this purpose except the `useradd` here. That would fail to create a user that already exists, and it seemed a shame to me to throw away the whole container's context because of one extra entry in `/etc/passwd`.

I'm happy to work with you on making sure the change is appropriate here (I love me a good dense shellism, but I understand why not everyone does :sweat_smile:), or to move it to a different PR if you think that'd be a better home for it. Just let me know!

Comment (jackh726) [reply src/ci/run.sh:11]: Another PR is better

**Comment by sethp** [comment]
> Some changes occurred in src/tools/cargo
>
> cc @ehuss

Sorry for the noise: there weren't any changes, I fumbled the submodule/rebase workflow. Should be backed out now.

**Comment by jackh726** [comment]
Can you squash your commits?

**Comment by sethp** [comment]
>  Can you squash your commits?

Sure, I did a little bit of editing down to cut out the stuff that's only useful for test usability. Let me know if you want me to do any more!

File: tests/ui/proc-macro/meta-macro-hygiene.rs [hunk 2]
```diff
@@ -3,8 +3,10 @@
 // edition:2018
 // compile-flags: -Z span-debug -Z macro-backtrace -Z unpretty=expanded,hygiene -Z trim-diagnostic-paths=no
 // check-pass
+// ignore-tidy-linelength
```
Comment (jackh726) [inline tests/ui/proc-macro/meta-macro-hygiene.rs:6]: Are these and similar changes unrelated to this PR? If so, can you move to separate PR?

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp [hunk 3]
```diff
@@ -177,11 +177,64 @@ extern "C" void LLVMTimeTraceProfilerFinish(const char* FileName) {
 GEN_SUBTARGETS
 #undef SUBTARGET
 
+// Thanks to https://gist.github.com/dabrahams/1528856
+namespace access_retrofit {
```
Comment (jackh726) [inline compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:181]: Is this change nice-to-have, or required? If it's just nice-to-have, I'd rather it be split into a separate PR, since I'm not personally comfortable reviewing this alone.

File: compiler/rustc_feature/src/active.rs [hunk 4]
```diff
@@ -315,6 +315,8 @@ declare_features! (
     (active, abi_msp430_interrupt, "1.16.0", Some(38487), None),
     /// Allows `extern "ptx-*" fn()`.
     (active, abi_ptx, "1.15.0", Some(38788), None),
+    /// Allows `extern "riscv-interrupt-m" fn()` and `extern "riscv-interrupt-s" fn()`.
+    (incomplete, abi_riscv_interrupt, "CURRENT_RUSTC_VERSION", Some(111889), None),
```
Comment (jackh726) [inline compiler/rustc_feature/src/active.rs:319]: What exactly makes the feature `incomplete`? Is there something in particular that doesn't work?

File: tests/ui/proc-macro/meta-macro-hygiene.rs (hunk 2 above)
Comment (sethp) [reply tests/ui/proc-macro/meta-macro-hygiene.rs:6]: The tests don't pass without these changes: there's more detail in https://github.com/rust-lang/rust/pull/111891/commits/4182a0e00e0e095ecbf598223001990c2777e1d5 but the upshot is that a previous change introduced an issue where these two files need to change any time a symbol is added. This is that change, and removes the coupling for the future.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (hunk 3 above)
Comment (sethp) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:181]: I wrote more about this in https://github.com/rust-lang/rust/pull/111891#issuecomment-1562122922 and https://github.com/rust-lang/rust/pull/111891/commits/7161c5e7869acb664dfae55e375504d336e0de87 : prior to this change, it was impossible to introduce riscv UI tests to the compiler; I'm not super confident either in whatever is happening here, but it does seem important to me to include in this context. Could we ask someone else to help us take a look here?

File: compiler/rustc_feature/src/active.rs (hunk 4 above)
Comment (sethp) [reply compiler/rustc_feature/src/active.rs:319]: Not to my knowledge, I wasn't sure what the right lifecycle for the feature was. I assumed it would start `incomplete` and then be marked `active` at some future point, once *waves hands* criteria were met. 

I read you to be saying that the criterion here is my own confidence in the feature, is that right? If that's the case, it sounds like I should mark it as "active," right?

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (hunk 3 above)
Comment (jackh726) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:181]: Let's tag @nikic and @rust-lang/wg-llvm 

It might still be worth to split this into a separate PR that can have a separate reviewer assigned.

File: compiler/rustc_feature/src/active.rs (hunk 4 above)
Comment (jackh726) [reply compiler/rustc_feature/src/active.rs:319]: Think of it this way: If this were to go insta-stable, is the feature ready? Does it have any known bugs/limitations, incomplete features, or unsoundness? If not, then it's probably not incomplete.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (hunk 3 above)
Comment (nikic) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:181]: This is obviously not acceptable. You probably no longer need this because LLVM 14 is no longer supported.

Comment (sethp) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:181]: Ah! Dropping LLVM 14 support may indeed decrease the urgency here, though I'm not totally comfortable abandoning it entirely. The side-effect-ful filter call will cause any architecture's UI tests to fail any time any supported version of LLVM doesn't recognize a feature that rustc knows about, thereby seeming to defeat the purpose of the feature filter (at least as far as UI tests go)?

File: compiler/rustc_feature/src/active.rs (hunk 4 above)
Comment (sethp) [reply compiler/rustc_feature/src/active.rs:319]: Thanks, that helps: there's no known limitations (outside what's already under thorough discussion on #3246), and it's in a similar boat to those other architectures re: soundness. I note those are all "active," too, so I'll go ahead and flag this one as such also.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (hunk 3 above)
Comment (jackh726) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp]: Thanks for removing this change here. If you feel strongly about this, a separate PR is likely the right way to go about this.

**Comment by jackh726** [comment]
Please remove the llvm changes/revert from the commit history, and squash the final test change commit into the previous ones. After that, I think this should be good to go.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp (hunk 3 above)
Comment (sethp) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp]: Well, the tests pass, so the good news is that means there's currently no features at the LLVM level that exist in any newer versions, which does indeed mean these changes can be decoupled. I did want to leave it up to @nikic what the resolution ought to be: I searched for a few hours before coming up with the approach here, and this was the only one I found that worked against system-provided LLVM. I'm open to an alternative approach, but if y'all would rather do nothing, then perhaps an issue documenting the problem would be a more fitting outcome than a PR with this change in it. What do you think?

Comment (jackh726) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp]: Opening an issue to document the problem is a reasonable approach to move forward with this.

**Comment by jackh726** [comment]
Thanks for your patience on this review!

@bors r+

**Comment by sethp** [comment]
Thanks @jackh726 for your help and guidance! **star wars voice** _almooooost there, stay on target_
